lib/spack/spack/environment.py

```diff
@@ -0,0 +1,200 @@
+import os
+import os.path
+import collections
+
+
+class AttributeHolder(object):
+    """
+    Policy that permits to store any kind of attribute on self. The attributes must be passed as key/value pairs
+    during the initialization of the instance.
+    """
+    def __init__(self, **kwargs):
```

Comment:
@tgamblin I thought of the `kwargs` as a way to attach attributes to commands. The idea was to make it possible to write something like:
```
env.set_env('A', value, who=spec, where=...)
```
in client code and being able to refer to the additional information from functions written by the same client.